Code/GraphMol/FileParsers/CMLWriter.cpp

```diff
@@ -0,0 +1,219 @@
+       atom_itr != atom_itr_end; ++atom_itr) {
+    const auto& atom = *atom_itr;
+    PRECONDITION(atom, "bad atom");
+    const auto src = atom->getIdx();
+    const auto* const mol = &atom->getOwningMol();
+
+    for (auto bond_itrs = mol->getAtomBonds(atom);
+         bond_itrs.first != bond_itrs.second; ++bond_itrs.first) {
+      auto* bptr = (*mol)[*bond_itrs.first];
+      auto* nptr = bptr->getOtherAtom(atom);
+      const auto dst = nptr->getIdx();
+      if (dst < src) {
+        continue;
+      }
+
+      auto& bond = bondArray.add("cml:bond", "");
+      bond.put("<xmlattr>.atomRefs2",
+               boost::format{"%1%%2% %1%%3%"} % atom_id_prefix % src % dst);
+
+      bond.put("<xmlattr>.id",
+               boost::format{"%1%%2%"} % bond_id_prefix % bond_id++);
+
+      const auto btype = bptr->getBondType();
+      switch (btype) {
+        case Bond::SINGLE:
+          bond.put("<xmlattr>.order", 1);
+          break;
+        case Bond::DOUBLE:
+          bond.put("<xmlattr>.order", 2);
+          break;
+        case Bond::TRIPLE:
+          bond.put("<xmlattr>.order", 3);
+          break;
+        case Bond::AROMATIC:
+          bond.put("<xmlattr>.order", 'A');
+          break;
+        case Bond::DATIVEONE:
+          FALLTHROUGH;
+        case Bond::DATIVE:
+          FALLTHROUGH;
+        case Bond::DATIVEL:
+          FALLTHROUGH;
+        case Bond::DATIVER:
+          bond.put("<xmlattr>.order", 1);
+          break;
+        // XXX RDKit extension: bond orders greater than 3
+        case Bond::QUADRUPLE:
+          bond.put("<xmlattr>.order", 4);
+          break;
+        case Bond::QUINTUPLE:
+          bond.put("<xmlattr>.order", 5);
+          break;
+        case Bond::HEXTUPLE:
+          bond.put("<xmlattr>.order", 6);
+          break;
+        // XXX RDKit extension: half-integer orders
+        case Bond::ONEANDAHALF:
+          bond.put("<xmlattr>.order", "1.5");
+          break;
+        case Bond::TWOANDAHALF:
+          bond.put("<xmlattr>.order", "2.5");
+          break;
+        case Bond::THREEANDAHALF:
+          bond.put("<xmlattr>.order", "3.5");
+          break;
+        case Bond::FOURANDAHALF:
+          bond.put("<xmlattr>.order", "4.5");
+          break;
+        case Bond::FIVEANDAHALF:
+          bond.put("<xmlattr>.order", "5.5");
+          break;
+        default:
+          bond.pop_back();
+      }
+    }
+  }
+
+  return pt;
+}
+}  // namespace
+
+std::string MolToCMLBlock(const ROMol& mol, int confId, bool kekulize) {
+  auto pt = molToPTree(mol, confId, kekulize);
+  if (pt.empty()) {
+    return "";
+  }
+  std::ostringstream ss;
+  boost::property_tree::write_xml(
+      ss, pt,
+      boost::property_tree::xml_writer_make_settings<std::string>(' ', 2));
+  return ss.str();
+}
+
+void MolToCMLFile(const ROMol& mol, const std::string& fName, int confId,
+                  bool kekulize) {
+  std::ofstream ofs{fName};
+  ofs << MolToCMLBlock(mol, confId, kekulize);
+}
+
```

Comment:
I would make three functions, one that writes to a stream

```
void MolToCMLBlock(ostream &stream, const ROMol &mol, int confId, bool kekulize);
```

And have the there two call this one

```
std::string MolToCMLBlock(const ROMol& mol, int confId, bool kekulize) {
  std::ostringstrream ss;
   MolToCMLBlock(ss, mol, confid, kekulize);
}

void MolToCMLFile(const ROMol& mol, const std::string& fName, int confId,
                   bool kekulize) {
   std::ofstream ofs{fName};
   MolToCMLBlock(ofsl, mol, confId, kekulize);
 }
```